src/pubsub/ua_pubsub_keystorage.c

```diff
@@ -509,6 +509,7 @@ sksClientCleanupCb(void *client, void *context) {
         sksClient->config.securityPoliciesSize = 0;
         sksClient->config.certificateVerification.context = NULL;
         sksClient->config.logger.context = NULL;
+        UA_free(sksClient->config.clientContext);
```

Comment:
The client context lifecycle should be controlled by the userland.
It could even be stack-allocated. And/or contain dynamically allocated memory.

Wouldn't it be possible that the shutdown of the SKS system is signaled in the configured callback so that user can handle the context lifecycle?

Reply:
In this case, the client context is used to store the data to call getSecurityKeys, in the callbacks when sks async connect is successful. It is created, used and destroyed only internally by the SKS client. So user will/shall not have access to it.